tests/components/zha/test_discover.py

```diff
@@ -30,45 +66,298 @@
+    """Test device discovery."""
+
+    with mock.patch(
+        "homeassistant.components.zha.core.channels.Channels.async_new_entity"
+    ) as new_ent:
+        channels = channels_mock(
+            device_info["endpoints"],
+            manufacturer=device_info["manufacturer"],
+            model=device_info["model"],
+            node_desc=device_info["node_descriptor"],
+        )
+
+    assert device_info["event_channels"] == sorted(
+        [ch.id for pool in channels.pools for ch in pool.relay_channels.values()]
+    )
+    assert new_ent.call_count == len(
+        [
+            device_info
+            for device_info in device_info["entity_map"].values()
+            if not device_info.get("default_match", False)
+        ]
+    )
+
+    for call_args in new_ent.call_args_list:
+        comp, ent_cls, unique_id, channels = call_args[0]
+        map_id = (comp, unique_id)
+        assert map_id in device_info["entity_map"]
+        entity_info = device_info["entity_map"][map_id]
+        assert set([ch.name for ch in channels]) == set(entity_info["channels"])
+        assert ent_cls.__name__ == entity_info["entity_class"]
+
+
+def _ch_mock(cluster):
+    """Return mock of a channel with a cluster."""
+    channel = mock.MagicMock()
+    type(channel).cluster = mock.PropertyMock(return_value=cluster(mock.MagicMock()))
+    return channel
+
+
+@mock.patch(
+    "homeassistant.components.zha.core.discovery.ProbeEndpoint"
+    ".handle_on_off_output_cluster_exception",
+    new=mock.MagicMock(),
+)
+@mock.patch(
+    "homeassistant.components.zha.core.discovery.ProbeEndpoint.probe_single_cluster"
+)
+def _test_single_input_cluster_device_class(probe_mock):
+    """Test SINGLE_INPUT_CLUSTER_DEVICE_CLASS matching by cluster id or class."""
+
+    door_ch = _ch_mock(zigpy.zcl.clusters.closures.DoorLock)
+    cover_ch = _ch_mock(zigpy.zcl.clusters.closures.WindowCovering)
+    multistate_ch = _ch_mock(zigpy.zcl.clusters.general.MultistateInput)
+
+    class QuirkedIAS(zigpy.quirks.CustomCluster, zigpy.zcl.clusters.security.IasZone):
+        pass
+
+    ias_ch = _ch_mock(QuirkedIAS)
+
+    class _Analog(zigpy.quirks.CustomCluster, zigpy.zcl.clusters.general.AnalogInput):
+        pass
+
+    analog_ch = _ch_mock(_Analog)
+
+    ch_pool = mock.MagicMock(spec_set=zha_channels.ChannelPool)
+    ch_pool.unclaimed_channels.return_value = [
+        door_ch,
+        cover_ch,
+        multistate_ch,
+        ias_ch,
+        analog_ch,
+    ]
+
+    disc.ProbeEndpoint().discover_by_cluster_id(ch_pool)
+    assert probe_mock.call_count == len(ch_pool.unclaimed_channels())
+    probes = (
+        (zha_const.LOCK, door_ch),
+        (zha_const.COVER, cover_ch),
+        (zha_const.SENSOR, multistate_ch),
+        (zha_const.BINARY_SENSOR, ias_ch),
+        (zha_const.SENSOR, analog_ch),
+    )
+    for call, details in zip(probe_mock.call_args_list, probes):
+        component, ch = details
+        assert call[0][0] == component
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5051600)